arch/rv32i/src/pmp.rs

```diff
@@ -47,19 +38,75 @@ register_bitfields![u8,
 ];
 
 /// Main PMP struct.
-pub struct PMP<const NUM_REGIONS: usize, const NUM_REGIONS_OVER_TWO: usize> {
+///
+/// Tock will ignore locked PMP regions. Note that Tock will not make any
+/// attempt to avoid access faults from locked regions.
+///
+/// `NUM_REGIONS_OVER_TWO`: The number of PMP regions divided by 2.
+///  The RISC-V spec mandates that there must be either 0, 16 or 64 PMP
+///  regions implemented. If you are using this PMP struct we are assuming
+///  there is more then 0 implemented. So this value should be either 8 or 32.
+///
+///  If however you know the exact number of PMP regions implemented by your
+///  platform and it's not going to change you can just specify the number.
+///  This means that Tock won't be able to dynamically handle more regions,
+///  but it will reduce runtime space requirements.
+///  Note: that this does not mean all PMP regions are connected.
+///  Some of the regions can be WARL (Write Any Read Legal). All this means
+///  is that accessing `NUM_REGIONS` won't cause a fault.
+pub struct PMP<const NUM_REGIONS_OVER_TWO: usize> {
     /// The application that the MPU was last configured for. Used (along with
     /// the `is_dirty` flag) to determine if MPU can skip writing the
     /// configuration to hardware.
     last_configured_for: MapCell<AppId>,
+    /// This is a 64-bit mask of locked regions.
+    /// Each bit that is set in this mask indicates that the region is locked
+    /// and cannot be used by Tock.
+    locked_region_mask: u64,
+    /// This is the total number of avaliable regions.
+    /// This will be between 0 and NUM_REGIONS_OVER_TWO * 2 depending
+    /// on the hardware and previous boot stages.
+    num_regions: usize,
 }
 
-impl<const NUM_REGIONS: usize, const NUM_REGIONS_OVER_TWO: usize>
-    PMP<NUM_REGIONS, NUM_REGIONS_OVER_TWO>
-{
-    pub const unsafe fn new() -> Self {
+impl<const NUM_REGIONS_OVER_TWO: usize> PMP<NUM_REGIONS_OVER_TWO> {
+    pub unsafe fn new() -> Self {
+        // RISC-V PMP can support from 0 to 64 PMP regions
+        // Let's figure out how many are supported.
+        // We count any regions that are locked as unsupported
+        let mut num_regions = 0;
+        let mut locked_region_mask = 0;
+
+        for i in 0..(NUM_REGIONS_OVER_TWO * 2) {
+            // Read the current value
+            let pmpcfg_og = csr::CSR.pmpcfg[i / 4].get();
+
+            // Flip R, W, X bits
+            let pmpcfg_new = pmpcfg_og ^ (3 << ((i % 4) * 8));
+            csr::CSR.pmpcfg[i / 4].set(pmpcfg_new);
+
+            // Check if the bits are set
+            let pmpcfg_check = csr::CSR.pmpcfg[i / 4].get();
+
+            if pmpcfg_check == pmpcfg_og {
+                // We didn't correctly flip all of the bits
+                if pmpcfg_og & ((1 << 7) << ((i % 4) * 8)) > 0 {
```

Comment:
This checks if the bit is locked

Reply:
Isn't that what the `if` on 91 does?

Reply:
There are two checks.

The first checks if all of the bits we set stuck. If they didn't stick it could be because the region is locked. or because the reason isn't implemented. So this second check tells if it's locked.

Reply:
I have updated the comments to make this clear